File: python/leetcode221.py
class Solution:
    def maximalSquare(self, matrix):
        """
        :type matrix: List[List[str]]
        :rtype: int
        """
        if len(matrix) == 0:
            return 0
        m, n = len(matrix), len(matrix[0])
        dp = [[0 if matrix[i][j] == "0" else 1 for j in range(n)] for i in range(m)]

        for i in range(1, m):
            for j in range(1, n):
                if dp[i][j] == 1:
                    dp[i][j] = min(dp[i - 1][j], dp[i][j - 1], dp[i - 1][j - 1]) + 1

        res = max([max(dp[i]) for i in range(m)])
        return res ** 2

File: leetcodec++/leetcode221.cpp
//
// Created by 黄敏辉 on 2019-09-05.
//

class Solution {
public:
    int maximalSquare(vector<vector<char>>& matrix) {

        int n = matrix.size();
        if(n == 0)
            return 0;
        int m = matrix[0].size();
        int dp[n+1][m+1];
        memset(dp, 0, sizeof(dp));
        int s = 0;

        for (int i = 1; i <= n; i++){
            for (int j = 1; j <= m; j++){
                if (matrix[i-1][j-1] == '0'){
                    dp[i][j] = 0;
                }
                else{
                    dp[i][j] = 1 + min(dp[i-1][j-1], min(dp[i-1][j], dp[i][j-1]));
                    s = max(s, dp[i][j]);
                }
            }
        }
        return s*s;
    }

};